clash-cores/src/Clash/Cores/Crc/Internal.hs

```diff
@@ -0,0 +1,704 @@
+          | crcB .&. _crcTopBitMask > 0 = (shiftL crcB 1) `xor` _crcpolyShifted
+          | otherwise                   = shiftL crcB 1
+        applyPoly crcW = L.iterate applyPolyBit crcW L.!! snatToNum _crcDataWidth
+        applyCrc crc w = applyPoly $ crc `xor` shiftL (extend w) (snatToNum _crcWidth)
+        nextCrc = applyCrc _crcCurrent $ applyWhen _crcReflectInput reverseBV word
+
+-- | Get current CRC state. This is the value before any output reflection and output XOR.
+getCrcState
+  :: forall (crcWidth :: Nat) (dataWidth :: Nat)
+   . SoftwareCrc crcWidth dataWidth
+  -> BitVector crcWidth
+getCrcState (SoftwareCrc {..}) = go _crcDataWidth _crcParams
+  where
+    go SNat (CrcParams SNat _ _ _ _ _) = fst $ split _crcCurrent
+
+-- | Generate the CRC
+--
+--   If the 'SoftwareCrc' will be reused, 'reset' must be called before starting
+--   to process the next message.
+digest
+  :: forall (crcWidth :: Nat) (dataWidth :: Nat)
+   . SoftwareCrc crcWidth dataWidth
+  -> BitVector crcWidth
+digest params@(SoftwareCrc {..}) = go _crcDataWidth _crcParams
+  where
+    go SNat (CrcParams SNat _ _ _ _ _) = outputCrcRev `xor` _crcXorOutput
+      where
+        CrcParams {..} = _crcParams
+        outputCrcRev = applyWhen _crcReflectOutput reverseBV $ getCrcState params
+
+-- | Compute the raw residue value for these 'CrcParams', which is the value left in the
+--   CRC register after processing any valid codeword. Raw in this case means
+--   this is the residue before the output reflection and out xor has been applied.
+--
+--   A fixed amount of padding can be given. This is useful during CRC verification
+--   in the scenario where where you want to handle more words in a single cycle
+--   than are native to the protocol.
+--
+--   For example ethernet is byte-oriented but you may want to handle 32 bit
+--   in a single cycle to achaive a high throughput.
+--   This means that it is possible for 8, 16, 24 or 32 bits in the last data
+--   word of the stream to be valid.
+--   We can calculate a residue for each case and compare it to the CRC result.
+--   This saves us from having to instantiate four seperate CRCs for the receive path.
+--   The invalid words must be set to zero.
+rawResidue
+  :: forall (crc :: Type)
+   . KnownCrc crc
+  => crc
+  -- ^ The CRC to use
+  -> Int
+  -- ^ How many invalid(zero) bits the stream is padded with
+  -> BitVector (CrcWidth crc)
+  -- ^ The residue
+rawResidue crc nExtra = rawResi
+  where
+    dataWidth = d1
+    CrcParams {..} = crcParams crc
+    swCrc = mkSoftwareCrc crc dataWidth
+    crcVal = applyWhen _crcReflectOutput reverse $ bv2v $ digest swCrc
+    bitsToFeed = fmap (v2bv . reverse . singleton) $
+                    V.toList crcVal L.++ L.replicate nExtra 0
+    rawResi = getCrcState $ L.foldl' feed swCrc bitsToFeed
+
+-- | Compute the residue value for these 'CrcParams', which is the value left in the
+--   CRC register after processing any valid codeword.
```

Comment:
```suggestion
--   CRC register after processing any valid codeword.
--
-- __NB__: This function is not synthesizable
```

Reply:
Ah, I opened the docs and saw that these functions are grouped under _Software_. That explains why they aren't annotated non-synthesisable. The thing is, these functions (or rather a compile-time evaluated version of them through Template Haskell) are really useful in hardware, so people will want to use them anyway. So let's document it.

```suggestion
--   CRC register after processing any valid codeword.
--
--   To use the output of this function as a constant in your design, you can
--   lift it through Template Haskell:
--
-- @
-- import qualified Data.List as L
--
-- res :: BitVector 32
-- res = $(lift $ rawResidue Crc32_ethernet 0)
--
-- ress :: Vec 4 (BitVector 32)
-- ress = $(listToVecTH $ L.map (rawResidue Crc32_ethernet) [0..3])
-- @
--
--   __NB__: This function is not synthesizable
--
```